Source/Common/ArgumentLoader.cpp

```diff
@@ -5,47 +5,250 @@
 #include <FEXCore/fextl/string.h>
 #include <FEXCore/fextl/vector.h>
 
-#include "cpp-optparse/OptionParser.h"
 #include "git_version.h"
 
-#include <stdint.h>
-
 namespace FEX::ArgLoader {
-void FEX::ArgLoader::ArgLoader::Load() {
-  RemainingArgs.clear();
-  ProgramArguments.clear();
-  if (Type == LoadType::WITHOUT_FEXLOADER_PARSER) {
-    LoadWithoutArguments();
-    return;
+enum class ParserTypes : uint32_t {
+  String,
+  BoolTrue,
+  BoolFalse,
+  StringEnum,
+  StringArray,
+  Max,
+};
+
+struct OptionDetails {
+  std::string_view Arg;
+  FEXCore::Config::ConfigOption Option;
+  ParserTypes ParseType;
+};
+
+constexpr static OptionDetails ArgToOption[] = {
+#define ARG_TO_CONFIG
+#include <FEXCore/Config/ConfigOptions.inl>
+};
+
+const OptionDetails* FindOption(std::string_view Argument) {
+  for (auto& Arg : ArgToOption) {
+    if (Arg.Arg == Argument) {
+      return &Arg;
+    }
   }
 
-  optparse::OptionParser Parser {};
-  Parser.version("FEX-Emu (" GIT_DESCRIBE_STRING ") ");
-  optparse::OptionGroup CPUGroup(Parser, "CPU Core options");
-  optparse::OptionGroup EmulationGroup(Parser, "Emulation options");
-  optparse::OptionGroup DebugGroup(Parser, "Debug options");
-  optparse::OptionGroup HacksGroup(Parser, "Hacks options");
-  optparse::OptionGroup MiscGroup(Parser, "Miscellaneous options");
-  optparse::OptionGroup LoggingGroup(Parser, "Logging options");
+  return nullptr;
+}
+
+void PrintHelp() {
+  const char* Arguments[3] {};
+  Arguments[0] = "man";
+  Arguments[1] = "FEX";
+  Arguments[2] = nullptr;
+
+  execvp("man", (char* const*)Arguments);
+}
 
-#define BEFORE_PARSE
+void ExitWithError(std::string_view Error) {
+  fextl::fmt::print("Error: {}\n", Error);
+  _exit(1);
+}
+
+class ArgParser final {
+public:
+  ArgParser(FEX::ArgLoader::ArgLoader* Loader)
+    : Loader {Loader} {}
+
+  void Parse(int argc, char** argv);
+  void Version(std::string_view version) {
+    _Version = version;
+  }
+
+  fextl::vector<fextl::string> GetGuestArgs() {
+    return std::move(RemainingArgs);
+  }
+  fextl::vector<const char*> GetParsedArgs() {
+    return std::move(ProgramArguments);
+  }
```

Comment:
If we need a way to pull all state underneath this class, the way to do it is to enforce move semantics:
```cpp
auto [Guest, Parsed] = std::move(arg_parser).GetArgs();
```

This is achieved by merging the two functions and making it callable with rvalues, only:
```cpp
std::pair<vector, vector> GetArgs() && {
    return make_pair(std::move(Args1), std::move(Args2));
}
```

Equivalently, the parsed arguments don't need to be state in ArgParser but could instead be returned from Parse() directly.

Reply:
Resolved.